Doc/library/__main__.rst

```diff
@@ -1,25 +1,102 @@
-
-:mod:`__main__` --- Top-level script environment
-================================================
+:mod:`__main__` --- CLIs, import-time behavior, and ``__name__ == '__main__'``
+==============================================================================
 
 .. module:: __main__
-   :synopsis: The environment where the top-level script is run.
+   :synopsis: CLIs, import-time behavior, and ``__name__ == '__main__'``
 
 --------------
 
-``'__main__'`` is the name of the scope in which top-level code executes.
-A module's __name__ is set equal to ``'__main__'`` when read from
-standard input, a script, or from an interactive prompt.
+In Python, ``__main__`` is not a single mechanism in the language, but in fact
+is part of two quite different constructs:
+
+1. The ``__name__ == '__main__'`` statement
+2. The ``__main__.py`` file in a Python packages
+
+Each of these mechanisms are related to Python :ref:`tut-modules`; both how
+users interact with them as well as how they interact with each other. See
+:ref:`tut-modules` for details.
+
+
+``__name__ == '__main__'``
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``'__main__'`` is the name of the environment where top-level code is run.
+"Top-level code" means when a Python module is initialized from an interactive
+prompt, from standard input, from a file argument, from a :option:`-c` argument
+or from a :option:`-m` argument, but **not** when it is initialized from an
+import statement.  In any of these situations, the module's ``__name__`` is set
+equal to ``'__main__'``.  The only other context in which Python code is run is
+when it is imported through an import statement. In that case, ``__name__`` is
+set equal to the module's name; usually the name of the file without the
+``.py`` extension.
+
+As a result, a module can discover whether or not it is running in the
+top-level environment by checking its own ``__name__``, which allows a common
+idiom for conditionally executing code when the module is not initialized from
+an import statement::
+
+    if __name__ == '__main__':
+        # Execute when the module is not initialized from an import statement.
```

Comment:
I use a pattern like:

```
if __name__ == "__main__":
    sys.exit(main())
```

this way the module is compatible with console script entrypoints

Reply:
I think it's worth mentioning that `if __name__ == "__main__"` should contain as few statements as possible - especially assignments as they have significantly different behaviour to assignments elsewhere

Reply:
Specifically when I use `if __name__ == "__main__":` it's always the penultimate line and the consequent is always `sys.exit(main()`

Reply:
@graingert  I added a section called "design patterns" which hits on these points (see 2b5f7104d75a2f4dd91256072ba30324bd403a91). The only thing I'm not sure about is the section title 'design patterns' itself. What do you think?

Reply:
Could be `Common patterns` or `Extra tips`

Reply:
I landed on "Idiomatic Usage." There are a few subsections with that title which go into details like these.